### ISSUE TITLE
Blocking Storage API fails with Bad address when used from Batch

### ISSUE THREAD

**Issue by natarajmb** [Azure/azure-sdk-for-java#5721]
**Query/Question**
I'm using the Blob Storage Service from Azure Batch which is called as a step from Data Factory. I have built and tested the Batch service (job) running locally and it works as expected. But when the job is triggered as part of the ADF step I'm seeing an issue with blobClient.downloadFile.

I'm using the blocking API from 12.0.0-Preview.3 

***Why is this not a Bug or a feature Request?***
I'm not sure if I'm using the API incorrectly or it behaves differently in the Batch environment. 

**Setup (please complete the following information if applicable):**
 - OS: Azure Batch
 - IDE : N/A
 - Version of the Library used - 12.0.0-Preview.3
 
Error seen on the Batch log.
AWKSplitter.java (101) - blobClient.downloadToFile(/path/to/a/file/created/beforehand)
```
Exception in thread "main" reactor.core.Exceptions$ReactiveException: java.io.IOException: Bad address
	at reactor.core.Exceptions.propagate(Exceptions.java:326)
	at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:91)
	at reactor.core.publisher.Mono.block(Mono.java:1494)
	at com.azure.storage.common.Utility.blockWithOptionalTimeout(Utility.java:241)
	at com.azure.storage.blob.BlobClient.downloadToFile(BlobClient.java:447)
	at com.azure.storage.blob.BlobClient.downloadToFile(BlobClient.java:411)
	at com.cns.demo.azure.batch.AWKSplitter.main(AWKSplitter.java:101)
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:93)
		... 5 more
Caused by: java.io.IOException: Bad address
	at java.base/sun.nio.ch.FileDispatcherImpl.pwrite0(Native Method)
	at java.base/sun.nio.ch.FileDispatcherImpl.pwrite(FileDispatcherImpl.java:68)
	at java.base/sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:109)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:58)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:50)
	at java.base/sun.nio.ch.SimpleAsynchronousFileChannelImpl$3.run(SimpleAsynchronousFileChannelImpl.java:374)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by rickle-msft** [comment]
Hi, @natarajmb. Thank you for posting your question. Can you please share some more information about the code you're running (preferably a snippet, but a description may suffice)? And do you have any logs from the SDK that you can share?

**Comment by rickle-msft** [comment]
After looking online a bit, there's not a ton of helpful information about this Bad Address exception, but what I can find seems to indicate that it can be thrown when the file system gets into a certain undesirable state (though I couldn't discern exactly what that is). Are there any patterns of file access/manipulation related to this Batch that might mess with the filesystem in strange ways? Like a file getting deleted while it's being read or opening tons of file handles or something to that effect?

**Comment by natarajmb** [comment]
Below is the code snippet I'm executing. It's a POC I'm trying to build using the Data Factory and Batch. Error quoted in the first the description is from the Batch console. Nothing much gets logged apart from once I pasted above.

The file I'm downloading is about 110MB XML file which I split down the line, but the since the first exception is thrown from the `blobClient.downloadToFile(OUTPUT_LOCATION + INPUT_FILE);` I think Batch process stops and exits with error code.

```
ContainerClient containerClient = new AWKSplitter().containerClient(batchURL, connectionString, container);
        BlobClient blobClient = containerClient
                .getBlobClient(folderPath + "/" + fileName);
        if (Files.exists(Paths.get(OUTPUT_LOCATION))) {
            try {
                Files.list(Paths.get(OUTPUT_LOCATION)).forEach(path -> {
                    try {
                        Files.delete(path);
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                });
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else {
            try {
                Files.createDirectories(Paths.get(OUTPUT_LOCATION));
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        blobClient.downloadToFile(OUTPUT_LOCATION + INPUT_FILE);
```

**Comment by rickle-msft** [comment]
@natarajmb Because this is working locally and because the root cause of the exception seems to be in the filesystem apis--something related to file manipulation, we do not now believe this is a problem in the SDK itself. For that reason, we won't be able to offer substantial support on this particular issue. Should you hit a problem in your POC caused by the sdk or the service, we will be more than happy to help.

**Comment by mkozioro** [comment]
Hi @rickle-msft 
I'm having exactly same problem. I'm getting it randomly when downloading same file.
Even if download process goes to the end without exception, file is broken (has bad content).
I don't see that problem with small files (couple MB). I'm receiving it when downloading bigger blob ~128MB. 

I'm creating client in following way:
```
    private static BlobContainerClient buildBlobContainerClient(final AzureQueueProperties azureQueueProperties) {
        String connectionString = requireNonNull(azureQueueProperties.getConnectionString(),
                "Azure blob connection string cannot be null.");
        return new BlobContainerClientBuilder().connectionString(connectionString)
                .containerName(azureQueueProperties.getBackupContainerName()).buildClient();
    }
```
and using created BlobContainerClient to build blockBlobClient:
```
this.blockBlobClient = requireNonNull(blobContainerClient).getBlobClient(queueBackupBlobName);
```
and downloading the file:
```
// Delete if file already exists
FileUtils.deleteQuietly(randomFileNamePath.toFile());
blockBlobClient.downloadToFile(randomFileNamePath.toString());
```
Blob was uploaded using same SDKs using:
```
blockBlobClient.uploadFromFile(serializedBatchInputMessages.getPath());
```

I'm using 12.0.0-preview.4 version of SDKs.

**Comment by rickle-msft** [comment]
@jianghaolu @alzimmermsft Can you guys take a look at this. This issue seems to be pretty rare, but I found a [similar issue](https://github.com/oracle/graal/issues/400#issuecomment-409976928) that seemed to be [related to the use of DirectByteBuffers](https://github.com/oracle/graal/issues/400#issuecomment-410895212). Maybe you guys have some insights there?

That said, this case seems to be coming from the AsynchronousFileChannel and not the ByteBuffers. I have a similar suspicion as in the original post that there is some sort of race or bad environment condition that leads to the file getting deleted mid write and hence trying to write to a bad address. I'm not very familiar with Azure Batch, so while this seems a little strange, it could be why these processes tend to work locally and not in Batch. I know we require the file to not exist before downloading to it, but for the sake of testing, could you try randomly generating a file name and writing to a file you know is unique and see if there's an improvement after getting rid of the delete calls? 

@alzimmermsft Perhaps we should add an overwrite flag to downloadToFile so customers don't have to worry about deleting files in advance.

**Comment by mkozioro** [comment]
Hi @rickle-msft 
In my scenario new SDKs were not running on Azure Batch env :). It's running on Apache Storm 1.2.3 cluster using newest OpenJDK 11.

**Comment by rickle-msft** [comment]
@natarajmb What JDK are you running on?

**Comment by natarajmb** [comment]
@rickle-msft I'm running on Oracle JDK 8 on local and OpenJDK on Ubuntu on the Batch box.

**Comment by rickle-msft** [comment]
@natarajmb Thank you for that information. This seems to support my hypothesis that it's Linux/OpenJDK related. Let's wait for mkozioro to see if he is able to run and verify on a windows machine.

**Comment by joshfree** [comment]
@sima-zhu could you assist by setting up an Ubuntu 18.04 environment that you and possibly @rickle-msft could use to investigate?

**Comment by mkozioro** [comment]
Hi @rickle-msft. We don't have software that can run on Windows, so we can't check it.

**Comment by sima-zhu** [comment]
I have the simple code snippet run successfully on Ububtu 18.04 with openjdk 11.

The problem might cause by the batch env or the way batch invoke the blob opertion.

@natarajmb @mkozioro Could you share a little more details on how you setup the batch box?
What env and APIs do I need to use?

**Comment by rickle-msft** [comment]
@mkozioro I just merged in a change for the next release that will delete files if the download fails so you don't get left with these broken files on your machine.

In addition to Sima's questions, can one of you also share the number and size of files you are processing and if your batch job is introducing any parallelism? This will help us ensure our attempt at reproing is as accurate as possible. @natarajmb

**Comment by mkozioro** [comment]
@sima-zhu 
I don't have batch box. 
My environment is Ubuntu:
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.3 LTS"

With java:
openjdk version "11.0.4" 2019-07-16
OpenJDK Runtime Environment (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3)
OpenJDK 64-Bit Server VM (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3, mixed mode, sharing)

SDKs are running as a part of topology deployed on Apache Storm 1.2.3 cluster. 
We are using blob, storage (both preview and stable ones) apis there.
What details could help?

**Comment by sima-zhu** [comment]
I have tried similar code snippet on Azure batch. 

**Procedure:**
1. Create a pool with ubuntu 18.04 compute nodes.
2. jar file with code snippets of uploading and downloading 1GB file.
3. Create a job and task to run the jarfile.

**Results:**
All upload and download operation done successfully. Also it passed the integrity check. 

I am not able to reproduce the issue in Azure batch. I did not try on Azure Batch SDK. But the issue cannot reproduce in Azure Box env.
